### PR TITLE
Anonymous Web SDK connection to domain server

### PR DESCRIPTION
**Pull request by ctrlaltdavid**
`DomainServer` API additions:
- `connect(url)`: WebSocket URL for the Doman Server's WebRTC signaling WebSocket port, e.g., `ws://127.0.0.1:40102`.
- `disconnect()`
- `location`: Read-only property returning the current location URL.
- `state`: Read-only numeric property returning the current connection state.
- `stateToString(state)`: Returns a string representation of the given connection state.
- `DISCONNECTED` etc.: Read-only properties of connection state values.
- `refusalInfo`: Read-only property with description of connection `REFUSED` state.
- `errorInfo`: Read-only property with description of connection `ERROR` state.
- `onStateChanged`: Write-only property that sets a callback function to be called when the connection state changes.

Example Web app provided in `./example` directory.

Builds on top of PR https://github.com/vircadia/vircadia-web-sdk/pull/20.
Works in conjunction with Vircadia PR https://github.com/vircadia/vircadia/pull/1300.

Merge sequence:
- [x] PR https://github.com/vircadia/vircadia-web-sdk/pull/20.
- [x] CR, QA.
- [x] Merge.

#### QA

With the domain server running on localhost:
- Enable anonymous connections and restart.

In the directory where the PR files are checked out:
` `npm install` - To set up the the dev & test environment.
- `npm run lint` - Should complete with no errors.
- `npm run build` - Should complete with no errors.
- `npm run test` - All  tests should pass. 
    Run the tests against the domain server in https://github.com/vircadia/vircadia/pull/1300.
    Check that the domain-server.exe process doesn't crash and restart during the tests.
    Known issue: "Jest did not exit one second after the test run has completed." See: https://github.com/vircadia/vircadia-web-sdk/issues/21
- `npm run devdoc` : Docs should be created in `./docs/dev/` directory.
- `npm run sdkdoc` : Docs should be created in `./docs/sdk/` directory.
- Set up and run the example Web app per the README.md in the `./example` directory.
  Note: If anonymous connection isn't permitted, the app will show a "REFUSED" state.
  Known issue: State reported by `DomainServer` isn't "ERROR" if location used is invalid.

### CODE REVIEW

File: src/DomainServer.ts [hunk 1]
```diff
@@ -10,21 +10,291 @@
...
+    }
+
+    get errorInfo(): string {
+        return this.#_errorInfo;
+    }
 
-    get isConnected() {
-        return this.#_isConnected;
+    set onStateChanged(callback: OnStateChangedCallback) {
+        if (typeof callback === "function" || callback === null) {
```
[the remark this answers is not in the document]
Comment (ctrlaltdavid) [reply src/DomainServer.ts:171]: Per the JSDoc, the value values are a function or `null`: "Sets a single function to be called when the state of the domain server connection changes. Set to null to remove the callback."
A value of `undefined` would be an error on the user script's part, and this is caught with an error message logged and the property value set to ``null`.
The JSDoc does need a minor update, thought to explicitly show that `null` is a valid value.

File: src/DomainServer.ts [hunk 2]
```diff
@@ -10,21 +10,291 @@
...
+    //      current domain and stops the check-ins from being sent. The C++ never stops sending checkin-ins.
+
+    /*@sdkdoc
+     *  Called when the state of the domain server connection changes.
+     *  @callback DomainServer~onStateChangedCallback
+     *  @param {DomainServer.ConnectionState} state - The state of the domain server connection.
+     *  @param {string} info - Refusal or error information if the state is <code>REFUSAL</code> or <code>ERROR</code>.
+     */
+
+    static get DISCONNECTED(): ConnectionState {
```
[the remark this answers is not in the document]
Comment (ctrlaltdavid) [reply src/DomainServer.ts:87]: `DomainServer` is part of the public API. I envisage a JavaScript Web client doing things such as ...
```
if (DomainServer.state === DomainServer.CONNECTED) {
    ...
}
```

An alternative could be ...
```
if (DomainServer.state === DomainServer.states.CONNECTED) {
    ...
}
```
... though I don't know that this is an improvement. It seems a little less immediate?

Assignment client APIs may well have similar states.

File: src/domain/networking/DomainHandler.ts [hunk 3]
```diff
@@ -131,12 +137,15 @@ class DomainHandler {
...
-     *  Sets whether Interface is connected to the domain.
-     *  @param {boolean} isConnected - <code>true</code> if Interface is connected to the domain, <code>false</code> if it
+     *  Sets whether the client is connected to the domain.
+     *  @param {boolean} isConnected - <code>true</code> if the client is connected to the domain, <code>false</code> if it
      *      isn't.
+     *  @param {boolean} forceDisconnect - <code>true</code> if any partly open communications channels to the domain server
+     *      should be closed (e.g., if currently trying to connect), <code>false</code> if they need not be.
      */
-    setIsConnected(isConnected: boolean): void {
+    setIsConnected(isConnected: boolean, forceDisconnect = false): void {
```
[the remark this answers is not in the document]
Comment (ctrlaltdavid) [reply src/domain/networking/DomainHandler.ts:146]: Yes, `forceDisconnect` is an optional parameter of type `boolean` and default value of `false`.
If `forceDisconnect?: boolean` is used the parameter type becomes `boolean | undefined`, which is not wanted.
If `forceDisconnect? = false` is used the parameter type becomes `boolean | undefined` and you get a TypeScript error, "Parameter cannot have question mark and initializer".
... So, no change.

**Review by Misterblue — APPROVED** [review]
JS has made it confusing with both 'undefined' and 'null' but the code doesn't fail in either case and the documentation specifies what it should be.

**Review by digisomni — CHANGES_REQUESTED** [review]

File: src/domain/networking/DomainHandler.ts [hunk 4]
```diff
@@ -150,14 +159,43 @@ class DomainHandler {
...
+
+        // clear member variables that hold the connection state to a domain
+        this._uuid = new Uuid(Uuid.NULL);
+
+        // WEBRTC TODO: Address further C++ code.
+
+        // WEBRTC TODO: Should C++ clear _domainConnectionRefusals also?
+        this._domainConnectionRefusals.clear();  // Re-report any refusals if retry connecting to the same domain.
+
+        console.log("[networking] Disconnecting from domain server.");
```
[the remark this answers is not in the document]
Comment (ctrlaltdavid) [reply src/domain/networking/DomainHandler.ts:189]: Quite possibly.

File: tests/DomainServer.unit.test.js [hunk 5]
```diff
@@ -13,16 +13,133 @@ import DomainServer from "../src/DomainServer";
...
+    test("Error state and callback if no location specified to connect to", () => {
+        const error = jest.spyOn(console, "error").mockImplementation(() => { /* noop */ });
+        const domainServer = new DomainServer();
+        domainServer.onStateChanged = (state) => {
+            expect(state).toBe(DomainServer.ERROR);
+        };
+        expect(domainServer.state).toBe(DomainServer.DISCONNECTED);
+        expect(domainServer.errorInfo).toBe("");
+        expect(domainServer.refusalInfo).toBe("");
+        domainServer.connect("  ");
```
[the remark this answers is not in the document]
Comment (ctrlaltdavid) [reply tests/DomainServer.unit.test.js:135]: Doesn't matter. I'll make it 0 spaces so as not to cause confusion.

**Review by digisomni — APPROVED** [review]

**Comment by digisomni** [comment]
Test suites passed.

**Pull request merged**